File: Solved problems/kattis/Zagrade.cpp
// tags: implementation, bitmask, bitwise
#pragma GCC optimize("Ofast")
#include <bits/stdc++.h>
using namespace std;
#define rep(i,a,b) for(int i = a; i < b; ++i)
// -------------------------------
int main() {
    ios::sync_with_stdio(false); 
    cin.tie(0); cout.tie(0);
    string line; cin >> line;
    stack<int> s;
    vector<pair<int,int>> brackets;
    rep(i, 0, line.size()) {
        if (line[i] == '(') s.push(i);
        else if (line[i] == ')') {
            brackets.emplace_back(s.top(), i); // C++11
            s.pop();            
        }
    }
    int n = brackets.size();
    vector<bool> skip;
    set<string> expressions;
    rep(mask, 1, 1 << n) {
        skip.assign(line.size(), false);
        int j = 0;        
        for (auto& b : brackets) {
            if (mask & (1 << j)) {
                skip[b.first] = true;
                skip[b.second] = true;
            }
            ++j;
        }
        string exp = "";
        rep(i, 0, line.size()) {
            if (skip[i]) continue;
            exp += line[i];
        }
        expressions.insert(exp);
    }
    for (auto& exp : expressions) cout << exp << '\n';
    return 0;
}

File: Solved problems/kattis/Zagrade.py
# parte 1: acumulamos pares de paréntesis
line = input().strip()
stack = []
brackets = []
for i in range(len(line)):
    if line[i] == '(':
        stack.append(i)
    elif line[i] == ')':
        brackets.append((stack.pop(), i))
n = len(brackets)
# parte 2: generamos todas las expresiones e imprimimos
expressions = set()
for mask in range(1, 1 << n): # x << n == x * (2 ^ n)
    skip = [False] * len(line)
    for j in range(n):
        if (1 << j) & mask > 0:
            skip[brackets[j][0]] = True
            skip[brackets[j][1]] = True
    exp = ''
    for j, c in enumerate(line):
        if skip[j]: continue
        exp += c
    expressions.add(exp)
expressions = sorted(list(expressions))
for exp in expressions: print(exp)

